modules/ads_lib/src/main/java/com/google/api/ads/adwords/lib/utils/BatchJobHelperUtility.java

```diff
@@ -0,0 +1,135 @@
+
+  private static final String TAG_MUTATE_RESPONSE = "mutateResponse";
+  private static final String TAG_RVAL = "rval";
+
+  private final Supplier<Transformer> transformerSupplier;
+  private final XMLInputFactory xmlInputFactory;
+  private final XMLOutputFactory xmlOutputFactory;
+  private final XMLEventFactory xmlEventFactory;
+
+  @Inject
+  BatchJobHelperUtility(
+      Supplier<Transformer> transformerSupplier,
+      XMLInputFactory xmlInputFactory,
+      XMLOutputFactory xmlOutputFactory,
+      XMLEventFactory xmlEventFactory) {
+    this.transformerSupplier = transformerSupplier;
+    this.xmlInputFactory = xmlInputFactory;
+    this.xmlOutputFactory = xmlOutputFactory;
+    this.xmlEventFactory = xmlEventFactory;
+  }
+
+  /**
+   * Returns a {@code ByteSource} object that wraps the batch job response into chunks with
+   * pagination, which will be parsed by Axis/JAX-WS.
+   */
+  public ByteSource buildWrappedByteSource(URL url, int startIndex, int numberResults)
+      throws IOException, XMLStreamException, TransformerException {
+    Preconditions.checkArgument(startIndex >= 0, "startIndex must not be negative");
+    Preconditions.checkArgument(numberResults > 0, "numberResults must be positive");
+
+    StringWriter stringWriter = new StringWriter();
+
+    XMLEventReader xmlEventReader = xmlInputFactory.createXMLEventReader(url.openStream());
+    XMLEventWriter xmlEventWriter = xmlOutputFactory.createXMLEventWriter(stringWriter);
+
+    int index = 0;
+    boolean included = false;
+
+    while (xmlEventReader.hasNext()) {
+      XMLEvent event = xmlEventReader.nextEvent();
+
+      if (event.isStartElement()) {
+        String name = ((StartElement) event).getName().getLocalPart();
+        if (name.equals(TAG_MUTATE_RESPONSE)) {
+          // Write the <mutateResponse> tag.
+          xmlEventWriter.add(event);
+          continue;
+        } else if (name.equals(TAG_RVAL) && ++index > startIndex) {
+          // Included the eligible <rval> tags and sub tags by setting the 'included' flag.
+          included = true;
+        }
+      } else if (event.isEndElement()) {
+        String name = ((EndElement) event).getName().getLocalPart();
+        if (name.equals(TAG_MUTATE_RESPONSE)) {
+          // Write the </mutateResponse> tag.
+          xmlEventWriter.add(event);
+          break;
+        } else if (included && name.equals(TAG_RVAL) && index == startIndex + numberResults) {
+          // Write the last </rval> tag and append a </mutateResponse> tag, if this has written
+          // the specified number of results.
+          xmlEventWriter.add(event);
+          xmlEventWriter.add(xmlEventFactory.createEndElement("", "", TAG_MUTATE_RESPONSE));
+          break;
+        }
+      }
+
+      if (included) {
+        // Write the included tags and events.
+        xmlEventWriter.add(event);
+      }
+    }
+    xmlEventReader.close();
+    xmlEventWriter.close();
+
+    InputStream truncatedInputStream =
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *truncatedInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when Supplier.get throws an exception. Either a) close *truncatedInputStream* in a try-finally block or b) close the resource by declaring *truncatedInputStream* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).